payment_bitcoin/models/bitcoin.py

```diff
@@ -0,0 +1,177 @@
+
+    markup = fields.Float('Markup (%)')
+    unit = fields.Selection([('BTC', 'BTC'), ('mBTC', 'mBTC')],
+                            'Display Unit', default='mBTC')
+    valid_minutes = fields.Integer(
+        'Rate Valid For (Minutes)',
+        default=20,
+        help=
+        "after this minutes rate will be checked again for same amount")
+
+    @api.model
+    def get_rate(self, order_id=False, order_ref=False):
+        # function returns bitcoin rate and address
+        # for the order currency and total amount
+        sobjs = self.search([])
+        if len(sobjs) != 1:
+            return False
+        sobj = sobjs[0]
+
+        if order_id:
+            order_obj = self.env['sale.order'].sudo().browse(int(order_id))
+        elif order_ref:
+            orders = self.env['sale.order'].\
+                search([('name', '=', order_ref)])
+            if not orders:
+                return False
+            order_obj = orders[0]
+
+        currency = order_obj.pricelist_id.currency_id
+        amount_total = order_obj.amount_total
+
+        addr_ids = self.env['bitcoin.address'].\
+            search([('order_id', '=', order_obj.id)], limit=1)
+        if not addr_ids:
+            addr_ids = self.env['bitcoin.address'].\
+                search([('order_id', '=', False)], limit=1)
+            if not addr_ids:
+                _logger.error('No Bitcoin Address configured')
+                return False
+
+        valid_rate_exists = self.env['bitcoin.rate.line'].\
+            sudo().search([('order_id', '=', order_obj.id),
+                           ('currency_id', '=', currency.id),
+                           ('amount', '=', amount_total),
+                           ('create_date', '>=',
+                            (datetime.now()-relativedelta(minutes=sobj.valid_minutes)).
+                            strftime('%Y-%m-%d %H:%M:00'))])
+        if valid_rate_exists:
+            # Rate was looked up within valid time limit,
+            # so we are using the valid one
+            rate = valid_rate_exists[0].rate
+        else:
+            # Check for New Rate
+            url = sobj.url.replace('{CURRENCY}', currency.name).\
+                replace('{AMOUNT}', str(amount_total))
+            response = requests.get(url)
+            if response.status_code != 200:
+                _logger.error('can not find Bitcoin exchange rate')
+                return False
+
+            rate = float(response.content)
+            # rate lookup entry saves in logs
+            self.env['bitcoin.rate.line'].sudo().create(
+                {'rate_id': sobj.id,
+                 'rate': rate,
+                 'amount': amount_total,
+                 'currency_id': currency.id,
+                 'order_id': order_obj.id})
+        if addr_ids and rate:
+            addr_ids[0].sudo().write(
+                {'order_id': order_obj.id})
+            b_addr = addr_ids[0].name
+
```

Comment:
This is a personal preference (although also advisable in PEP8), but too much empty lines in between a method makes difficult the code reading